### PR TITLE
I725 more log priors

### PR DESCRIPTION
**Pull request by fcooper8472**
#725 adds inverse gamma and log normal log priors.

Also reimplements Cauchy, and orders unit tests alphabetically.

### CODE REVIEW

File: pints/_log_priors.py [hunk 1]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+        if self._a <= 0:
+            raise ValueError('Shape parameter a must be positive')
+        if self._b <= 0:
+            raise ValueError('Scale parameter b must be positive')
+
+        # Cache constants
+        self._k = self._a * np.log(self._b) - scipy.special.gammaln(self._a)
+        self._ap1 = self._a + 1.
+
+    def __call__(self, x):
```
Comment (ben18785) [inline pints/_log_priors.py]: Checked

File: pints/_log_priors.py [hunk 2]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+        _x = float(x[0])
+
+        if _x <= 0.0:
+            return -np.inf
+        else:
+            return self._k - self._ap1 * np.log(_x) - np.divide(self._b, _x)
+
+    def evaluateS1(self, x):
```
Comment (ben18785) [inline pints/_log_priors.py]: Checked!

File: pints/_log_priors.py [hunk 3]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+            return val, np.asarray([0.])
+        else:
+            return val, np.asarray(
+                [np.divide(self._b - self._ap1 * _x, _x * _x)])
+
+    def n_parameters(self):
+        """ See :meth:`LogPrior.n_parameters()`. """
+        return 1
+
+    def sample(self, n=1):
```
Comment (ben18785) [inline pints/_log_priors.py]: checked

File: pints/_log_priors.py [hunk 4]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+    .. math::
+        f(x|\text{logmean},\text{scale}) = \frac{1}{x\;\text{scale}\sqrt{2\pi}}
+        \exp\left(-\frac{(\log x-\text{logmean})^2}{2\;\text{scale}^2}\right)
+
+    and expectation
+
+    .. math::
+        \mathrm{E}(X)=\exp\left(\text{logmean}+\frac{\text{scale}^2}{2}\right).
+
+    For example, to create a prior with logmean of ``0`` and a scale of ``1``,
```
Comment (ben18785) [inline pints/_log_priors.py]: I'd add a line explaining what the log_mean is: "The log_mean parameter of a log-normal distribution is the mean of a normal distribution whose random samples, when exponented, yield a sample from a log-normal distribution".

File: pints/_log_priors.py [hunk 5]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+
+    def sample(self, n=1):
+        """ See :meth:`LogPrior.sample()`. """
+        return scipy.stats.invgamma.rvs(a=self._a, scale=self._b, loc=0.,
+                                        size=n)
+
+
+class LogNormalLogPrior(pints.LogPrior):
+    r"""
+    Defines a log-normal (log) prior with a given ``logmean`` and scale
```
Comment (ben18785) [inline pints/_log_priors.py]: Can we change logmean to log_mean as think it's good to separate the two words.

File: pints/_log_priors.py [hunk 6]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+            raise ValueError('Scale must be positive')
+
+        # Cache constants
+        self._offset = -np.log(self._scale * np.sqrt(2. * np.pi))
+        self._1on2sigsq = 1. / (2. * self._scale * self._scale)
+        self._m1onsigsq = -1. / (self._scale * self._scale)
+        self._sigsqmmu = self._scale * self._scale - self._logmean
+
+    def __call__(self, x):
+        if x[0] < 0.0:
```
Comment (ben18785) [inline pints/_log_priors.py]: Think this should <= as the Log-normal doesn't have support for 0 (sorry if this is incorrect in the distribution zoo app!)

File: pints/_log_priors.py [hunk 7]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+            return -np.inf
+        else:
```
Comment (ben18785) [inline pints/_log_priors.py]: checked!

File: pints/_log_priors.py [hunk 8]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+            _lx = np.log(x[0])
+            _shift = _lx - self._logmean
+            return self._offset - _lx - self._1on2sigsq * _shift * _shift
+
+    def evaluateS1(self, x):
```
Comment (ben18785) [inline pints/_log_priors.py]: checked!

File: pints/_log_priors.py [hunk 9]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+            _x = x[0]
+            _lx = np.log(_x)
+            return self(x), np.asarray(
+                [self._m1onsigsq * np.divide(self._sigsqmmu + _lx, _x)])
+
+    def n_parameters(self):
+        """ See :meth:`LogPrior.n_parameters()`. """
+        return 1
+
+    def sample(self, n=1):
```
Comment (ben18785) [inline pints/_log_priors.py]: Looks good!

File: pints/tests/test_log_priors.py [hunk 10]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+            to_test = [point]
+            self.assertAlmostEqual(
+                scipy.stats.gamma.logpdf(to_test[0], a=a1, scale=1. / b1),
+                p1(to_test), places=9)
+            self.assertAlmostEqual(
+                scipy.stats.gamma.logpdf(to_test[0], a=a2, scale=1. / b2),
+                p2(to_test), places=9)
+
+        # Test derivatives
+        p1_derivs = [0., -879.345, -11.115, -3.222, -2.70295918367347,
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Really nice!

File: pints/tests/test_log_priors.py [hunk 11]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+
+        for point, deriv in zip(points, p1_derivs):
+            calc_val, calc_deriv = p1.evaluateS1([point])
+            self.assertAlmostEqual(calc_deriv[0], deriv)
+
+        for point, deriv in zip(points, p2_derivs):
+            calc_val, calc_deriv = p2.evaluateS1([point])
+            self.assertAlmostEqual(calc_deriv[0], deriv)
+
+        # Test pathological edge case
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Haha, nice!

File: pints/tests/test_log_priors.py [hunk 12]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+        p1 = pints.GammaLogPrior(5.0, 0.25)
+        self.assertEqual(len(p1.sample()), 1)
+
+        n = 1000
+        samples1 = p1.sample(n)
+        self.assertEqual(len(samples1), n)
+
+        # Mean should be ~ 5/0.25 = 20, so we check that this is very roughly
+        # the case, but we can be very relaxed as we only check it's not ~1.25
+        mean = np.mean(samples1).item()
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Can always increase the sample size here to ensure we don't get false positives...

File: pints/tests/test_log_priors.py [hunk 13]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+        d = 1
+        n = 1
+        x = p.sample(n)
+        self.assertEqual(x.shape, (n, d))
+        n = 10
+        x = p.sample(n)
         self.assertEqual(x.shape, (n, d))
 
-        # Roughly check distribution (main checks are in numpy!)
+        # Very roughly check distribution (main checks are in numpy!)
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Haha, yep.

File: pints/_log_priors.py [hunk 14]
```diff
@@ -427,14 +432,145 @@ def sample(self, n=1):
...
+class LogNormalLogPrior(pints.LogPrior):
+    r"""
+    Defines a log-normal (log) prior with a given ``logmean`` and scale
+    ``scale``, with pdf
+
+    .. math::
+        f(x|\text{logmean},\text{scale}) = \frac{1}{x\;\text{scale}\sqrt{2\pi}}
+        \exp\left(-\frac{(\log x-\text{logmean})^2}{2\;\text{scale}^2}\right)
+
+    and expectation
```
Comment (ben18785) [inline pints/_log_priors.py]: I think it's worth creating mean functions for the log priors. I will make this a separate ticket. This makes the job of a modeller even easier since they just call this function to check the parameterisation...

File: pints/tests/test_log_priors.py [hunk 15]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+            to_test = [point]
+            self.assertAlmostEqual(
+                scipy.stats.invgamma.logpdf(to_test[0], a=a1, scale=b1),
+                p1(to_test), places=9)
+            self.assertAlmostEqual(
+                scipy.stats.invgamma.logpdf(to_test[0], a=a2, scale=b2),
+                p2(to_test), places=9)
+
+        # Test derivatives
+        p1_derivs = [0., np.inf, 2.343877e6, 223.27, 1.2220000000000002,
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Nice!

File: pints/tests/test_log_priors.py [hunk 16]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+                                                               scale=b1))
+
+        for point, deriv in zip(points, p2_derivs):
+            calc_val, calc_deriv = p2.evaluateS1([point])
+            self.assertAlmostEqual(calc_deriv[0], deriv)
+            self.assertAlmostEqual(calc_val,
+                                   scipy.stats.invgamma.logpdf(point, a=a2,
+                                                               scale=b2))
+
+    def test_inverse_gamma_prior_sampling(self):
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Ok, cool. Again could increase sample size if false positives are an issue.

File: pints/tests/test_log_priors.py [hunk 17]
```diff
@@ -154,122 +183,352 @@ def test_composed_prior_sampling(self):
...
+
+            pints_val_2 = p2([point])
+            scipy_val_2 = scipy.stats.lognorm.logpdf(point, scale=np.exp(mu2),
+                                                     s=sd2)
+
+            self.assertAlmostEqual(pints_val_1, scipy_val_1)
+            self.assertAlmostEqual(pints_val_2, scipy_val_2)
+
+        # Test derivatives
+        p1_derivs = [0., 278.54579293277163, -5.589063346695011,
```
Comment (ben18785) [inline pints/tests/test_log_priors.py]: Nice!

File: pints/tests/test_log_priors.py (hunk 12 above)
Comment (fcooper8472) [reply pints/tests/test_log_priors.py]: Yep, I'll leave for now.  Wonder whether there's a more robust way of doing it - should we be doing something like mean +- n*sd?

**Comment by fcooper8472** [comment]
@ben18785  I have made all those changes

**Comment by ben18785** [comment]
@MichaelClerx I'm happy to merge this whenever you are.

File: pints/_log_priors.py [hunk 18]
```diff
@@ -51,14 +52,14 @@ def __init__(self, a, b):
 
     def __call__(self, x):
         if x[0] < 0.0 or x[0] > 1.0:
-            return -float('inf')
+            return -np.inf
```
Comment (MichaelClerx) [inline pints/_log_priors.py]: Interesting https://stackoverflow.com/questions/42315541/difference-between-np-inf-and-floatinf

File: pints/_log_priors.py [hunk 19]
```diff
@@ -51,14 +52,14 @@ def __init__(self, a, b):
...
         else:
             return scipy.special.xlogy(self._a - 1.0,
                                        x[0]) + scipy.special.xlog1py(
                 self._b - 1.0, -x[0]) - self._log_beta
 
     def evaluateS1(self, x):
-        """ See :meth:`LogPDF.evaluateS1()`. """
+        r""" See :meth:`LogPDF.evaluateS1()`. """
```
Comment (MichaelClerx) [inline pints/_log_priors.py]: Please don't start doing this in places where we don't have maths!

File: pints/_log_priors.py [hunk 20]
```diff
@@ -628,7 +777,7 @@ def __init__(self, lower_or_boundaries, upper=None):
             self._value = 1
 
     def __call__(self, x):
-        return self._value if self._boundaries.check(x) else self._minf
+        return self._value if self._boundaries.check(x) else -np.inf
```
Comment (MichaelClerx) [inline pints/_log_priors.py]: Could be np.ninf apparently :p
